File: C-Abi-Bright-Aot/Tests/CAbiBridge.Crypto.Tests/TestCrypto/TestCryptoAesGcm/CryptoAesGcmFileStreamTest.cs


using System.Diagnostics;

namespace michele.natale.Tests;

using Pointers; 
using static CryptoTestUtils;

partial class CryptoAesGcmTest
{

  public async static Task TestAesGcmFileAsync(int rounds)
  {
    Console.Write($"{nameof(TestAesGcmFileAsync)}: ");

    string src = "data", dest = "cipher", srcr = "datar";

    var sw = Stopwatch.StartNew();
    for (var i = 0; i < rounds; i++)
    {
      var max = (1 << 21) + 1024;
      var flength = Random.Shared.Next(max);
      await SetRngFileDataAsync(src, flength);

      var associated = RngBytes(Random.Shared.Next(1, 64));
      using var key = new UsIPtr<byte>(RngBytes(NetServices.AES_GCM_MAX_KEY_SIZE));

      await NetServices.EncryptionFileAesGcmAsync(src, dest, key, associated);
      await NetServices.DecryptionFileAesGcmAsync(dest, srcr, key, associated);

      if (!NetServices.FileEquals(src, srcr))
        throw new Exception();

      if (i % (rounds / 10) == 0)
        Console.Write(".");
    }

    sw.Stop();

    var t = sw.ElapsedMilliseconds;
    Console.WriteLine($" rounds = {rounds}, t = {t}ms, td = {t / rounds}ms");
  }
}

File: C-Abi-Bridge-Aot/Tests/CAbiBridge.Aot.Cpp.Tests/crypto_aes_gcm_file_stream_test.cpp
#include "pch.h"

#include <vector>
#include <string> 
#include <fstream>
#include <chrono>
#include <iostream>

#include "bridge.h"
#include "variables.h"
#include "crypto_aes_gcm_test.h"


namespace michele::natale::Tests
{

  void test_aes_gcm_file(int rounds)
  {
    using namespace std::chrono;
    using namespace michele::natale::Cpp;

    std::cout << "test_aes_gcm_file_aot: ";

    const std::string src = "data";
    const std::string dest = "cipher";
    const std::string srcr = "datar";

    auto start = high_resolution_clock::now();


    int max = (1 << 21) + 1024;
    for (int i = 0; i < rounds; i++)
    {
      // Datei mit Zufallsgröße erzeugen 
      int flength = rng_int(1, max);
      set_rng_file_data(src, flength);

      // Associated Data
      auto associated = rng_bytes(rng_int(1, 64));

      // Key
      std::vector<uint8_t> key(Services::AES_GCM_MAX_KEY_SIZE);
      auto keydata = rng_bytes(key.size());
      std::copy(keydata.begin(), keydata.end(), key.begin());

      // Encrypt
      auto err = aes_gcm_encrypt_file_aot(
        (const uint8_t*)src.c_str(), (int)src.size(),
        (const uint8_t*)dest.c_str(), (int)dest.size(),
        key.data(), (int)key.size(),
        associated.data(), (int)associated.size());
      assert_error(err);

      // Decrypt
      err = aes_gcm_decrypt_file_aot(
        (const uint8_t*)dest.c_str(), (int)dest.size(),
        (const uint8_t*)srcr.c_str(), (int)srcr.size(),
        key.data(), (int)key.size(),
        associated.data(), (int)associated.size());
      assert_error(err);

      // Vergleich
      if (!equal_files_aot(
        reinterpret_cast<const uint8_t*>(src.data()), (int)src.length(),
        reinterpret_cast<const uint8_t*>(srcr.data()), (int)srcr.length(), &err))
        throw std::runtime_error("AES-GCM file mismatch");
      assert_error(err);

      if (i % (rounds / 10) == 0)
        std::cout << ".";
    }

    auto end = high_resolution_clock::now();
    auto ms = duration_cast<milliseconds>(end - start).count();

    std::cout << " rounds = " << rounds
      << ", t = " << ms << "ms"
      << ", td = " << (ms / rounds) << "ms\n";
  }



}